XiEditor/EditView.swift

```diff
@@ -233,14 +260,126 @@ class EditView: NSView {
         // store current event so that it can be sent to the core
         // if the selector for the event is "noop:".
         currentEvent = theEvent;
-        self.interpretKeyEvents([theEvent]);
+        self.inputContext?.handleEvent(theEvent);
         currentEvent = nil;
     }
-
-    override func insertText(insertString: AnyObject) {
-        sendRpcAsync("insert", params: insertedStringToJson(insertString as! NSString))
+    
+    // NSTextInputClient protocol
+    func insertText(aString: AnyObject, replacementRange: NSRange) {
+        self.removeMarkedText()
+        self.replaceCharactersInRange(replacementRange, withText: aString)
+//       sendRpcAsync("insert", params: insertedStringToJson(aString as! NSString))
     }
-
+    
+    func replacementMarkedRange(replacementRange: NSRange) -> NSRange {
+        var markedRange = _markedRange
+        
+        
+        if (markedRange.location == NSNotFound) {
+            markedRange = _selectedRange
+        }
+        if (replacementRange.location != NSNotFound) {
+            var newRange: NSRange = markedRange
+            newRange.location += replacementRange.location
+            newRange.length += replacementRange.length
+            if (NSMaxRange(newRange) <= NSMaxRange(markedRange)) {
+                markedRange = newRange
+            }
+        }
+        
+        return markedRange
+    }
+    
+    func replaceCharactersInRange(aRange: NSRange, withText aString: AnyObject) -> NSRange {
+        var replacementRange = aRange
+        var len = 0
+        if let attrStr = aString as? NSAttributedString {
+            len = attrStr.string.characters.count
+        } else if let str = aString as? NSString {
+            len = str.length
+        }
+        if (replacementRange.location == NSNotFound) {
+            replacementRange.location = 0
+            replacementRange.length = 0
+        }
+        for var i in 0..<aRange.length {
+            sendRpcAsync("delete_backward", params: [])
```

Comment:
This is probably the part of the PR I'm having the most trouble with. Right now it should work most of the time, but that will change. The issue is that NSRange counts UTF-16 code units, where "delete_backward" currently does one codepoint, but that's a placeholder for more sophisticated logic. Even now it wouldn't be right for non-BMP characters.

I think the best way to fix this is to have an RPC for exactly "replaceCharactersInRange", and pass in a precise count. Currently counts passed over the wire are all utf-8, so the front-end has to do the conversion, but I'm open to changing that.

There are other more subtle issues with perhaps the cursor position not aligning because of concurrent edits (say, by a plugin). That I'm prepared to deal with but will likely require some rocket science (like passing revision id's across the wire and doing transforms on the edit operations to effectively "rebase" them onto a potentially newer revision).

More generally, I'm willing to take your patch now, and refine it later, either myself or as an open issue, because I think having something working is more important than having it perfect.